pkg/util/blockdevice/blockdevice_linux.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package blockdevice
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"k8s.io/klog"
+)
+
+// findBlockDeviceRescanPath Find the underlaying disk for a linked path such as /dev/disk/by-path/XXXX or /dev/mapper/XXXX
+// will return /sys/devices/pci0000:00/0000:00:15.0/0000:03:00.0/host0/target0:0:1/0:0:1:0/rescan
+func findBlockDeviceRescanPath(path string) (string, error) {
+	devicePath, err := filepath.EvalSymlinks(path)
+	if err != nil {
+		return "", err
+	}
+	// if path /dev/hdX split into "", "dev", "hdX" then we will
+	// return just the last part
+	parts := strings.Split(devicePath, "/")
+	if len(parts) == 3 && strings.HasPrefix(parts[1], "dev") {
+		return filepath.EvalSymlinks(filepath.Join("/sys/block", parts[2], "device", "rescan"))
+	}
+	return "", fmt.Errorf("illegal path for device " + devicePath)
+}
+
+// getBlockDeviceSize returns the size of the block device by path
+func getBlockDeviceSize(path string) (int64, error) {
+	fd, err := os.Open(path)
+	if err != nil {
+		return 0, err
+	}
+	defer fd.Close()
+	pos, err := fd.Seek(0, io.SeekEnd)
+	if err != nil {
+		return 0, fmt.Errorf("error seeking to end of %s: %s", path, err)
+	}
+	return pos, nil
+}
+
+func checkBlockDeviceSize(devicePath string, deviceMountPath string, newSize int64) error {
+	klog.V(4).Infof("Detecting %q volume size", deviceMountPath)
+	size, err := getBlockDeviceSize(devicePath)
+	if err != nil {
+		return err
+	}
+
+	klog.V(3).Infof("Detected %q volume size: %d", deviceMountPath, size)
+
+	if size < newSize {
+		return fmt.Errorf("current volume size is less than expected one: %d < %d", size, newSize)
+	}
+
+	return nil
+}
+
+func RescanBlockDeviceGeometry(devicePath string, deviceMountPath string, newSize int64) error {
+	if newSize == 0 {
+		klog.Error("newSize is empty, skipping the block device rescan")
+		return nil
+	}
+
+	// when block device size corresponds expectations, return nil
+	bdSizeErr := checkBlockDeviceSize(devicePath, deviceMountPath, newSize)
+	if bdSizeErr == nil {
+		return nil
+	}
```

Comment:
if  getBlockDeviceSize fails , with this code you are executing the rescan, which might not be required? Is it the same latest code after all comments addressed as I remember i have given same comment before?

Reply:
Your comments were related to another code. If the size cannot be detected, then we can rely on further `/rescan` call try to detect the volume size once again.